core/controller/src/main/scala/whisk/core/controller/Meta.scala

```diff
@@ -34,13 +38,110 @@ import whisk.core.entity.types._
 import whisk.http.ErrorResponse.terminate
 import whisk.http.Messages
 import whisk.core.WhiskConfig
+import whisk.utils.JsHelpers._
 
-/**
- * A singleton object which defines the properties that must be present in a configuration
- * in order to implement the Meta API.
- */
-object WhiskMetaApi {
+private case class Context(
+    method: HttpMethod,
+    headers: List[HttpHeader],
+    path: String,
+    query: Map[String, String],
+    body: Option[JsObject] = None) {
+    val requestParams = query.toJson.asJsObject.fields ++ { body.map(_.fields) getOrElse Map() }
+    val overrides = WhiskMetaApi.reservedProperties.intersect(requestParams.keySet)
+    def withBody(b: Option[JsObject]) = Context(method, headers, path, query, b)
+    def metadata(user: Option[Identity]): Map[String, JsValue] = {
+        Map("__ow_meta_verb" -> method.value.toLowerCase.toJson,
+            "__ow_meta_headers" -> headers.map(h => h.name -> h.value).toMap.toJson,
```

Comment:
Should we normalize the `name` to be lowercased as well? There is `lowercasedName` on `HttpHeader`